### DISCUSSION TITLE
Change of keys on Bitemporal Tables

### DISCUSSION THREAD

**Discussion by Flastica** [hettie-d/pg_bitemporal#11] [General]
Hello @hettie-d ,
I'm starting to use and play with bitemporal models on my database, starting from your repo..
First of all... thanks for the amazing work you and your team have done, it really is something!

Now I'm having a problem that I don't know whether to label it as "structural" or design related.
I have the necessity to enable edits on columns of a bitemporal table, and for all operations (add, edit, delete)  I hit a wall.

I wondered if you had some thoughts on the matter or even if you think it should/shouldn't be possibile to do such things.

# Add:
What about "historical" integrity? An old row should be considered changed since it has new column/values, or since its changes are structural it shouldn't be traced?

# edit:
Historical Integrety after a label change?, 
With type/length changes it seems tricky... to me, the only way to maintain old data seems  a new column entirely, and to map externally which one is the 'active' column.

# delete:
Historical integrity by maintaining the old column, but mapping it externally as inactive?

For keys it seems even more NOPE scenario.
# Add: 
Adding a key, with existing data, would mark the necessity to add a default for the new key on old rows (what about integrity again?)+ the same problems as a normal add.
# Edit/Delete -> brain stack overflow
    
    
Have you any thought on the matter? Do you evaluate this is as a "limit" o a bitemporal structure or simply an error in design principle that should avoid changes on bitemporal tables?

**Comment by hettie-d** [comment]
Hi @Flastica, thank you so much for your comments! Here is what I've been doing so far with the structural changes. In all the cases, it "worked for me," meaning that for the tasks I needed to accomplish, it worked. There might be cases when my solutions don't work for others, and I am happy to brainstorm together and accept contributions. 

1) For adding a column, I just added new columns as to normal tables, and for all "before" records, the value in these columns would be null. 
2). I do not support historical renaming, so if I rename, I rename the whole column; I didn't have any use case for doing otherwise
3). Changing the column type is also done for the whole table; I never had cases where there would be any business value in knowing when the type changed, but in general, it might be worth having a "historical pg_catalog," and I am open to this idea.
4) I never had a business case for column deletion, possibly because when you work with bitemporal data, everybody's mindset changes :), but you can create a trigger on insert which would NULL the value of a column you want to "bitemporarily delete") 
5)  I am not sure what you mean by adding a key. You can have more than one  bitemporarily unique columns in one table, and same as with "normal" tables, there is no real difference between the business key and the bt-unique column, both conditions will be validated on insert. 

Happy to continue this conversation! Regards, Hettie D.

**Reply by Flastica** [reply]
Thank you for your answer!
I agree that this all depends a lot on the environment you are using it in.

1. About column deletion:
Since in my domain I already have a middle stage in which columns are validated and memorized (structurally) elsewhere, in my first thoughts I didn't consider to manage the whole aspect database-wise but more through client-based code. This being said, it seems logic to have the library manage it all by itself and leave the client out of it. I'll give it some thought!
2. Simple and efficient idea with a trigger! This leaves the client "free" and it solves the problem. Ideally tho I'm thinking of a way to signal explicitly when a given column doesn't exist anymore, rather than it having "only" null values. Maybe I'm over-reaching as database logics, but I'll do some digging.

3. These are a bit edge cases, but in my domain they still happen so I'll give some more detail:
- Add key: 
**After** a table is created and some record are inserted, the user realizes that another key column should be added for future records. 
   - Normal tables: If there are no conflicts, this is possible by adding a DEFAULT or similar for the new column without a sweat.
   - Bitemporal : Adding a key doesn't cause problems per se, but the DEFAULT does mess a bit with historical integrity one again.
   About this I can only come up with terrible ideas such as:
    -> the new key is treated as a standard column, and record are validated client-side when doing every operation (TERRIBLE)
    -> the key is added, and default is applied on old rows, but when querying the data, the old records are "cleaned" of the default value (ALSO TERRIBLE)
   
- Delete key: 
(Edit: While writing this I recognize that this case still maintain integrity and safe data, but I'll leave it for the sake of the conversation)
Premise: This will be likely NOT possible client-side, but I was wondering how to give even more consistency and integrity  at table data. 
Considering tables with more than one key, an user deletes a key column and it doesn't cause conflicts:
   - "No conflits" should be further investigated. For example:
   
        | id1 |  id2  | effective | 
        |-|-|-|
        | 1 | 2 | [2012-01-01 .., 2014-01-01..) |
        | 2 | 2 | [2020-01-01 .., infinity) |
     
     At first the two rows are two entirely different records.
     After id1 deletion, data is still ok integrity wise with no conflicts, but now the 2 rows are two versions of the same record.
     
     While reading old rows, if user doesn't know that id1 was a key, it's not clear that he's reading two different records.
     
Do you think that managing these cases on the database side can offer benefits as for instance more integrity, or exclusively more complexity for its own sake?

Best Regards,
Francesco

**Comment by hettie-d** [comment]
Hi Francesco, I think my problem here is that I am unsure what you mean by the "key" exactly. There are no "keys" in relational theory. We have columns (eq to attributes), and we have primary keys, unique constraints, and foreign key constraints. If you would elaborate on what a "key" means for you, I would be able to answer your question better. Thanks!

**Comment by Flastica** [comment]
Since on bitemporal tables primary keys are set with a single column _uid, in my examples I was referring to a logic unique key, for which an exclude constraint is created already.
For instance, for the add case the exclude constraint would look like:

..USING GIST(old_column with =, new_column with =, _effective with &&, _asserted with &&)

Sorry for being vague before.
I hope this clarify it now.

**Reply by hettie-d** [reply]
OK, that's a "business key." And what I was saying before was that you can add any number of bitemporarly unique constraints to a table. UNIQUNEss allows nulls, as always. However, if you want to completely remove a column that used to be a business key for this table, it sort of defeats the purpose of bitemporality, in my opinion. Technically speaking, you can make the old business key nullable, but you can't make a new business key NOT NULL if it didn't exist in the past. If it did exist and was not null all along, you should be able to swap. At least, that's a great theorem to try to prove! If you run some experiments with it and get some good results, let me know! Also, I hope you had a chance to go through the presentations that are out there in the docs subdirectory. (Which reminds me that I should add the more recent presentation!)

**Reply by Flastica** [reply]
Oh damn.
I was so fixed on the concept on business keys on standard tables, where they usually are also primary keys, that I didn't realize that for bitemporal tables we have unique keys.... even if I wrote "unique" in my last answer😪

This makes it a bit lighter, and I will try for sure some experiments!!
I'll keep you updated if interested!

I'm gonna go back to the presentation and re-read what I could've missed. 
Thanks a lot!

Francesco
